### ISSUE TITLE
expr.py seems to have a memory issue?

### ISSUE THREAD

**Issue by jfy4** [lehner/gpt#129]
I am having trouble with some code where the memory keeps growing.  In my script I have run `tracemalloc` to try to find the issue and it seems to be line 347 of expr.py, where it runs `cgpt.eval`.  below is a minimal working example where I see the problem:

```
import tracemalloc
tracemalloc.start()
import argparse
import gpt as g
import itertools as it

parser = argparse.ArgumentParser()
parser.add_argument('--mpi', default="1.1.1.1", type=str, help='mpi ranks passed to gpt')
globals().update(vars(parser.parse_args()))

# build lattice
L = 4
grid = g.grid([L,L,L,L], g.double)

# build identity gauge field
lnU = [ g.mspin(grid) for mu in range(0,4) ]
for mu in range(0,4):
    lnU[mu][:] = 0
g.message("\ninitialized gauge field to identity, lnU_{0}[0,0,0,0] = ", lnU[0][0,0,0,0])

# build Spin(4) generators
Ji2 = [ [ (g.gamma[a].tensor()*g.gamma[b].tensor() - g.gamma[b].tensor()*g.gamma[a].tensor())/8 for b in range(0,4) ] for a in range(0,4) ]
g.message("\nbuilt Spin(4) generators")
for a in range(0, 4):
    for b in range(0, 4):
        g.message("i/2 * J_{",a,b,"} = ", Ji2[a][b])

# build random spin connection
rng = g.random("seed string")
omega = [ [ [ rng.normal(g.complex(grid)) for b in range(0,4) ] for a in range(0,4) ] for mu in range(0, 4) ]
#for mu in range(0, 4):
#    for a in range(0, 4):
#        for b in range(0, a):
#            omega[mu][b][a] = g.eval( omega[mu][a][b]*(-1) )
g.message("\nbuilt random spin connection")
for mu in range(0, 4):
    for a in range(0, 4):
        for b in range(0, 4):
            g.message("omega_{",mu,"}^{",a,b,"}[0,0,0,0] = ", omega[mu][a][b][0,0,0,0])

# build gauge field from matrix exponential
for mu in range(0, 4):
    for a in range(0, 4):
        for b in range(0, 4):
            lnU[mu] += Ji2[a][b]*omega[mu][a][b]
g.message("(i/2 * J_{ab} * omega_0^{ab})[0,0,0,0] = ", lnU[0][0,0,0,0])
U = [ g.mspin(grid) for mu in range(0,4) ]
for mu in range(0,4):
    U[mu] = g.matrix.exp(lnU[mu])
g.message("\nU_mu = exp(i/2 * J_{ab} * omega_mu^{ab})")
g.message("U[0][0,0,0,0] = ", U[0][0,0,0,0])

# test matrix exponential function
tol = 1/1e12
test1 = g.eval(g.gamma["I"].tensor()*g.component.cos(omega[0][0][0]) + g.gamma[1].tensor()*g.component.sin(omega[0][0][0])*1j)[0,0,0,0]
test2 = g.eval(g.matrix.exp(g.gamma[1].tensor()*omega[0][0][0]*1j))[0,0,0,0]
eps2 = g.norm2(test1-test2)
g.message("\ntesting matrix exponentiation ||a-b||^2=", eps2,"=0?")
assert(eps2 < tol**2)

# test gauge field is unitary
g.message("\ntesting unitarity")
for mu in range(0,4):
    one = g.complex(grid)
    one[:] = 1
    eye = g.eval(g.gamma["I"].tensor()*one)
    eps = U[mu]*g.adj(U[mu]) - eye
    eps2 = g.norm2(eps)
    g.message("|U_{",mu,"} * U^dagger_{",mu,"} - 1|^2/Vol^2 = ",eps2/L**8,"=0?")
    assert(eps2/L**8 < tol**2)

################################################
# added by Judah trying to find memory leak ####    
for _ in range(40000):
    Gmunu = g.qcd.gauge.field_strength(U, 0, 1)
snapshot = tracemalloc.take_snapshot()
top_stats = snapshot.statistics('lineno')

print("[ Top 10 ]")
for stat in top_stats[:10]:
    print(stat)
assert False
#################################################
```

If you do 
```
while True:
    Gmunu = g.qcd.gauge.field_strength(U, 0, 1)
```
instead you can see the memory grow indefinitely. 
Let me know if I'm being dumb.  Thanks.

**Comment by daknuett** [comment]
```python
import gpt as g

L = [8,8,8,16]
grid = g.grid(L, g.double)
U = g.qcd.gauge.unit(grid)

while(True):
    Gmunu = g.qcd.gauge.field_strength(U, 0, 1)
    with open("/proc/self/statm") as fin:
        g.message(f"RSS (pages): {fin.read().split(' ')[1]}")

```

reproduces this issue. I can confirm that `gpt/core/expr.py:347` is the problem.

**Comment by daknuett** [comment]
I can confirm that the memory leak we experienced when training ML models or using them as preconditioners also (at least partially) originates from `gpt/core/expr.py:347`.

**Comment by lehner** [comment]
Should be fixed in commit 5da478a .
